File: kernel/coc/store.py
"""Campaign storage under <workspace>/.coc (contract §3). Every write is atomic."""

from __future__ import annotations

import datetime as _dt
import re
from pathlib import Path
from typing import Any

from .errors import RpcError
from .fileio import (append_jsonl, canonical_json, read_json, read_jsonl, sha256_text,
                     write_json_atomic)

CAMPAIGN_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")
CALL_ID = re.compile(r"^t(?P<turn>\d+)-c(?P<n>\d+)$")


def now_iso() -> str:
    return _dt.datetime.now(_dt.timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def parse_call_id(call_id: Any) -> tuple[int, int]:
    if not isinstance(call_id, str) or not CALL_ID.match(call_id):
        raise RpcError("invalid_params", "call_id must look like t<turn>-c<n>",
                       fix="the extension mints call_id; the model never writes it",
                       details={"call_id": call_id})
    match = CALL_ID.match(call_id)
    assert match is not None
    return int(match.group("turn")), int(match.group("n"))


def params_digest(params: dict[str, Any]) -> str:
    return sha256_text(canonical_json(params))


def fresh_turn(number: int, state: str = "awaiting_player",
               pending_choice: dict[str, Any] | None = None) -> dict[str, Any]:
    return {
        "turn": number,
        "state": state,
        "player_text": None,
        "opened_at": now_iso(),
        "calls": {},
        "receipts": [],
        "pending_choice": pending_choice,
    }


class Store:
    def __init__(self, workspace: Path) -> None:
        self.workspace = Path(workspace)
        self.root = self.workspace / ".coc"
        self.campaigns_dir = self.root / "campaigns"
        self.repos_dir = self.root / "repos"
        #: §14.1: the module store, shared by every campaign of the same module.
        self.modules_dir = self.root / "modules"

    def campaign_ids(self) -> list[str]:
        if not self.campaigns_dir.exists():
            return []
        return sorted(p.name for p in self.campaigns_dir.iterdir()
                      if (p / "campaign.json").exists())

    def campaign_dir(self, campaign_id: str) -> Path:
        return self.campaigns_dir / campaign_id

    def repo_dir(self, campaign_id: str) -> Path:
        return self.repos_dir / f"{campaign_id}.git"

    def validate_new_id(self, campaign_id: Any) -> str:
        if not isinstance(campaign_id, str) or not CAMPAIGN_ID.match(campaign_id):
            raise RpcError("invalid_params", "campaign id must be a short slug",
                           fix="use letters, digits, '.', '_' or '-' (max 64 chars)")
        if self.campaign_dir(campaign_id).exists():
            raise RpcError("invalid_params", f"campaign {campaign_id!r} already exists",
                           fix="pick another id or open the existing campaign")
        return campaign_id

    def open(self, campaign_id: Any, *, require_turn: bool = True) -> "Campaign":
        """`require_turn=False` is for `table.open`, which may rebuild a lost turn.json
        from the continuation checkpoint before anything reads it (contract §12.2)."""
        if not isinstance(campaign_id, str) or not campaign_id:
            raise RpcError("invalid_params", "params.campaign is required")
        campaign = Campaign(self, campaign_id)
        if not campaign.campaign_json.exists():
            raise RpcError("campaign_not_found", f"no campaign {campaign_id!r}",
                           fix="call campaign.list, or campaign.create",
                           details={"campaigns": self.campaign_ids()})
        required = [campaign.world_json] + ([campaign.turn_json] if require_turn else [])
        for path in required:
            if not path.exists():
                raise RpcError("campaign_not_ready",
                               f"campaign {campaign_id!r} is missing {path.name}")
        return campaign


class Campaign:
    def __init__(self, store: Store, campaign_id: str) -> None:
        self.store = store
        self.id = campaign_id
        self.dir = store.campaign_dir(campaign_id)
        self.repo_dir = store.repo_dir(campaign_id)
        self.campaign_json = self.dir / "campaign.json"
        self.world_json = self.dir / "world.json"
        self.turn_json = self.dir / "turn.json"
        self.party_dir = self.dir / "party"
        self.turns_dir = self.dir / "turns"
        self.transcript_path = self.dir / "transcript.jsonl"
        self.events_path = self.dir / "events.jsonl"
        self.telemetry_path = self.dir / "telemetry.jsonl"
        # slice 2 (contract §3, §12): the rebuildable checkpoint and the memory stores
        self.checkpoint_path = self.dir / "save" / "continuation" / "latest.json"
        self.memory_dir = self.dir / "memory"
        self.episodes_path = self.memory_dir / "episodes.jsonl"
        self.candidates_path = self.memory_dir / "candidates.jsonl"
        self.jobs_dir = self.memory_dir / "jobs"
        self.backlog_path = self.memory_dir / "backlog.jsonl"
        #: §14.8: a text handout is materialized here so the attachment is a file path.
        self.handouts_dir = self.dir / "handouts"

    # ---- documents --------------------------------------------------------

    def read_campaign(self) -> dict[str, Any]:
        return read_json(self.campaign_json)

    def write_campaign(self, data: dict[str, Any]) -> None:
        write_json_atomic(self.campaign_json, data)

    def read_world(self) -> dict[str, Any]:
        return read_json(self.world_json)

    def write_world(self, data: dict[str, Any]) -> None:
        write_json_atomic(self.world_json, data)

    def read_turn(self) -> dict[str, Any]:
        return read_json(self.turn_json)

    def write_turn(self, data: dict[str, Any]) -> None:
        write_json_atomic(self.turn_json, data)

    def party(self) -> list[dict[str, Any]]:
        if not self.party_dir.exists():
            return []
        return [read_json(p) for p in sorted(self.party_dir.glob("*.json"))]

    def write_sheet(self, sheet: dict[str, Any]) -> None:
        write_json_atomic(self.party_dir / f"{sheet['id']}.json", sheet)

    # ---- closed turns -----------------------------------------------------

    def turn_record_path(self, number: int) -> Path:
        return self.turns_dir / f"{number:04d}.json"

    def read_turn_record(self, number: int) -> dict[str, Any] | None:
        path = self.turn_record_path(number)
        return read_json(path) if path.exists() else None

    def write_turn_record(self, record: dict[str, Any]) -> None:
        write_json_atomic(self.turn_record_path(int(record["turn"])), record)

    def closed_turns(self) -> list[dict[str, Any]]:
        if not self.turns_dir.exists():
            return []
        return [read_json(p) for p in sorted(self.turns_dir.glob("*.json"))]

    # ---- append-only logs -------------------------------------------------

    def append_transcript(self, turn: int, role: str, text: str) -> dict[str, Any]:
        entry = {"turn": turn, "role": role, "text": text, "at": now_iso()}
        append_jsonl(self.transcript_path, entry)
        return entry

    def read_transcript(self) -> list[dict[str, Any]]:
        return read_jsonl(self.transcript_path)

    def read_events(self) -> list[dict[str, Any]]:
        return read_jsonl(self.events_path)

    def append_telemetry(self, row: dict[str, Any]) -> None:
        """One flat line per lane step (contract §8, §12.5). The extension writes its own
        rows to the same file; the kernel's carry `lane`."""
        append_jsonl(self.telemetry_path, {"at": now_iso(), **row})

    def turn_records_by_number(self) -> dict[int, dict[str, Any]]:
        return {int(r["turn"]): r for r in self.closed_turns()}

    # ---- idempotency (contract §2) ----------------------------------------

    def lookup_call(self, turn: dict[str, Any], call_id: str) -> dict[str, Any] | None:
        stored = (turn.get("calls") or {}).get(call_id)
        if stored is None:
            call_turn, _ = parse_call_id(call_id)
            record = self.read_turn_record(call_turn)
            stored = ((record or {}).get("calls") or {}).get(call_id)
        return stored

    def replay_or_conflict(self, turn: dict[str, Any], call_id: str,
                           params: dict[str, Any]) -> dict[str, Any] | None:
        """Return the stored result (marked replayed) for a repeated call_id, raise on
        a different payload, or None when the call is new."""
        digest = params_digest(params)
        stored = self.lookup_call(turn, call_id)
        if stored is None:
            return None
        if stored.get("params_sha256") != digest:
            raise RpcError("idempotency_conflict",
                           f"call_id {call_id} was already used with different params",
                           fix="mint a new call_id for a new call",
                           details={"call_id": call_id})
        return {**stored["result"], "replayed": True}

    @staticmethod
    def remember_call(turn: dict[str, Any], call_id: str, params: dict[str, Any],
                      result: dict[str, Any]) -> None:
        turn.setdefault("calls", {})[call_id] = {
            "params_sha256": params_digest(params),
            "result": result,
            "at": now_iso(),
        }

File: kernel-ts/write/store.ts
/** Campaign writes and remembered calls; each RPC reloads authoritative disk state. */
import { readFile, unlink } from 'node:fs/promises';
import { join, relative, resolve, sep } from 'node:path';
import type { KernelContext } from '../context.js';
import type { JsonObject, JsonValue, ReadonlyJson } from '../json.js';
import { jsonDigest, isJsonObject } from '../json.js';
import { appendJsonl, writeJsonAtomic } from '../fileio.js';
import { RpcError } from '../errors.js';
import type { CampaignWritePort, DomainEvent, ResolveCommit, TurnTransaction, WriteMethod, WriteStart } from '../transactions.js';
import { clone, array, row, number, string, repr, type Row } from '../read/values.js';
export const nowIso = (): string => new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
export const freshTurn = (turn: number, state = 'awaiting_player', pending: Row | null = null): Row => ({
    turn,
    state,
    player_text: null,
    opened_at: nowIso(),
    calls: {},
    receipts: [],
    pending_choice: pending,
});
export const missingContribution = (name: string): never => {
    throw new RpcError('not_implemented', `The ${name} contribution is not implemented in the TypeScript transaction runtime`);
};
export function required(params: Row, key: string, optional = false): string | null {
    const value = params[key];
    if (value == null) {
        if (optional)
            return null;
        throw new RpcError('invalid_params', `params.${key} must be a non-empty string`);
    }
    if (typeof value !== 'string' || !value.trim())
        throw new RpcError('invalid_params', `params.${key} must be a non-empty string`);
    return value;
}
export function parseCallId(value: any): [
    number,
    number
] {
    const match = typeof value === 'string' ? /^t(\d+)-c(\d+)$/.exec(value) : null;
    if (!match)
        throw new RpcError('invalid_params', 'call_id must look like t<turn>-c<n>', {
            fix: 'the extension mints call_id; the model never writes it',
            details: {
                call_id: value ?? null
            },
        });
    return [Number(match[1]), Number(match[2])];
}
export function turnStateError(turn: Row, method: string, allowed: string): RpcError {
    return new RpcError('turn_state', `${method} is not allowed while the turn is ${repr(turn.state)}`, {
        fix: allowed,
        details: {
            turn: turn.turn,
            state: turn.state
        },
    });
}
export const EVENT_TYPES: ReadonlySet<string> = new Set(['turn-started', 'player-declared', 'roll-resolved', 'scene-moved', 'clue-discovered', 'time-advanced',
    'turn-finalized', 'resource-changed', 'decision-settled', 'session-changed', 'choice-asked', 'memory-written', 'setup-completed',
    'handout-shown', 'item-transferred', 'definition-created', 'definition-queued', 'ability-acquired', 'flag-set', 'note-written', 'ruling-made',
    'npc-changed', 'dossier-established', 'journal-written', 'worldline-forked', 'worldline-switched', 'worldline-merged', 'adaptation-accepted',
    'turn-stranded']);
export class CampaignWriter implements CampaignWritePort {
    readonly directory: string;
    constructor(readonly context: KernelContext, readonly id: string) { this.directory = join(context.campaignsRoot, id); }
    path(name: string): string { return join(this.directory, name); }
    async read(name: string): Promise<Row> { return clone(row(await this.context.snapshots.readJson(this.path(name)))); }
    write(name: string, value: Row): Promise<void> { return writeJsonAtomic(this.path(name), value); }
    readCampaign() { return this.read('campaign.json'); }
    readWorld() { return this.read('world.json'); }
    readTurn() { return this.read('turn.json'); }
    private savePath(name: string): string {
        const root = join(this.directory, 'save'), target = resolve(root, name), path = relative(root, target);
        if (!name || !path || path === '..' || path.startsWith('..' + sep))
            throw new RpcError('invalid_params', 'Save document must remain inside its campaign save directory');
        return target;
    }
    async readSave(name: string): Promise<JsonValue | null> {
        const path = this.savePath(name);
        return await this.context.snapshots.pathExists(path) ? clone(await this.context.snapshots.readJson(path)) as JsonValue : null;
    }
    writeSave(name: string, value: ReadonlyJson): Promise<void> { return writeJsonAtomic(this.savePath(name), value); }
    async removeSave(name: string): Promise<void> {
        try { await unlink(this.savePath(name)); }
        catch(error){ if((error as NodeJS.ErrnoException).code!=='ENOENT')throw error; }
    }
    saveExists(name: string): Promise<boolean> { return this.context.snapshots.pathExists(this.savePath(name)); }
    saveDirectories(name: string): Promise<string[]> {
        return this.context.snapshots.sortedChildNames(this.savePath(name), path => this.context.snapshots.isDirectory(path));
    }
    writeCampaign(value: Row) { return this.write('campaign.json', value); }
    writeWorld(value: Row) { return this.write('world.json', value); }
    writeTurn(value: Row) { return this.write('turn.json', value); }
    writeSheet(value: Row) { return this.write(join('party', `${value.id}.json`), value); }
    recordName(turn: number): string { return join('turns', `${String(turn).padStart(4, '0')}.json`); }
    async readTurnRecord(turn: number): Promise<Row | null> {
        const path = this.recordName(turn);
        return await this.context.snapshots.pathExists(this.path(path)) ? this.read(path) : null;
    }
    writeTurnRecord(value: Row) { return this.write(this.recordName(number(value.turn)), value); }
    async files(folder: string): Promise<Row[]> {
        const names = await this.context.snapshots.sortedChildNames(this.path(folder), path => this.context.snapshots.isFile(path));
        return Promise.all(names.filter(name => name.endsWith('.json')).map(name => this.read(join(folder, name))));
    }
    party() { return this.files('party'); }
    records() { return this.files('turns'); }
    async appendTranscript(turn: number, role: 'player' | 'keeper', text: string): Promise<Row> {
        const entry = {
            turn,
            role,
            text,
            at: nowIso()
        };
        await appendJsonl(this.path('transcript.jsonl'), entry);
        return entry;
    }
    async appendEvent(turn: number, event: DomainEvent, callId?: string): Promise<void> {
        if (!EVENT_TYPES.has(event.type))
            throw new Error(`unknown event type ${event.type}`);
        let count = 0;
        try {
            count = (await readFile(this.path('events.jsonl'), 'utf8')).split('\n').filter(line => line.trim()).length;
        }
        catch (error) {
            if ((error as NodeJS.ErrnoException).code !== 'ENOENT')
                throw error;
        }
        await appendJsonl(this.path('events.jsonl'), {
            seq: count + 1,
            turn,
            type: event.type,
            at: nowIso(),
            ...(callId != null ? {
                call_id: callId
            } : {}),
            ...(event.receipt != null ? {
                receipt: event.receipt
            } : {}),
            data: event.data
        });
    }
    telemetry(value: Row) {
        return appendJsonl(this.path('telemetry.jsonl'), {
            at: nowIso(),
            ...value
        });
    }
    async replay(turn: Row, callId: string, params: Row): Promise<Row | null> {
        let stored = row(turn.calls)[callId];
        if (stored == null)
            stored = row((await this.readTurnRecord(parseCallId(callId)[0]))?.calls)[callId];
        if (stored == null)
            return null;
        if (stored.params_sha256 !== jsonDigest(params))
            throw new RpcError('idempotency_conflict', `call_id ${callId} was already used with different params`, {
                fix: 'mint a new call_id for a new call',
                details: {
                    call_id: callId
                },
            });
        return {
            ...stored.result,
            replayed: true
        };
    }
}
export function rememberCall(turn: Row, callId: string, params: Row, result: Row): void {
    turn.calls ??= {};
    turn.calls[callId] = {
        params_sha256: jsonDigest(params),
        result,
        at: nowIso()
    };
}
export function createTurnTransaction(campaign: CampaignWriter, world: Row, turn: Row): TurnTransaction {
    return {
        campaign,
        world,
        turn,
        async beginWrite(method: WriteMethod, params: JsonObject, options = {}): Promise<WriteStart> {
            const [, ordinal] = parseCallId(params.call_id), callId = string(params.call_id);
            const replay = await campaign.replay(turn, callId, params);
            if (replay)
                return {
                    kind: 'replay',
                    result: replay
                };
            const meta = await campaign.readCampaign();
            if (meta.status === 'completed' && method !== 'table.ask' && method !== 'table.narrate') {
                const action = row(params.action), effects = array(params.effects), requested = string(action.decision || '').trim();
                const decision = requested.startsWith('decision:') ? requested : `decision:coc7:${requested}`;
                const accounting = method === 'table.resolve' && action.intent === 'montage'
                    && Object.keys(action).every(key => ['intent', 'goal', 'method', 'decision', 'ending', 'actor', 'scenario_san_reward_expr'].includes(key))
                    && ['decision:coc7:development:end-session', 'decision:coc7:development:settle-ending'].includes(decision);
                const correction = method === 'table.apply' && !Object.hasOwn(row(meta.ending), 'scope') && effects.length === 1
                    && isJsonObject(effects[0]) && effects[0].kind === 'ending' && effects[0].scope === 'chapter'
                    && Object.keys(effects[0]).every(key => ['kind', 'scope', 'summary'].includes(key));
                if (!accounting && !correction)
                    throw new RpcError('campaign_not_ready', 'this campaign is completed; only late development accounting is writable', {
                        fix: 'keep the ending intact; use explicit development:end-session or pending development:settle-ending with intent montage, then narrate',
                    });
            }
            const opening = options.allowOpening && number(turn.turn) === 0 && turn.state === 'awaiting_player';
            if (!['open', 'acting'].includes(turn.state) && !opening)
                throw turnStateError(turn, method, 'wait for player_input to open a turn');
            return {
                kind: 'new',
                callId,
                ordinal
            };
        },
        async touchActing() {
            if (turn.state === 'open') {
                turn.state = 'acting';
                await campaign.writeTurn(turn);
            }
        },
        async commitResolve(commit: ResolveCommit) {
            turn.receipts.push(...commit.receipts);
            turn.state = 'acting';
            rememberCall(turn, commit.callId, commit.params, commit.result);
            await campaign.writeTurn(turn);
            for (const event of commit.events)
                await campaign.appendEvent(number(turn.turn), event, commit.callId);
        },
    };
}
